plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/procedure/TestIcebergAddFilesProcedure.java

```diff
@@ -0,0 +1,547 @@
+
+        assertQueryFails(
+                "CALL iceberg.system.add_files('tpch', '" + icebergTableName + "', 'tpch', '" + hiveTableName + "')",
+                "Expected target 'string' type, but got source 'int' type");
+
+        assertUpdate("DROP TABLE hive.tpch." + hiveTableName);
+        assertUpdate("DROP TABLE iceberg.tpch." + icebergTableName);
+    }
+
+    @Test
+    void testAddFilesDifferentDataColumnDefinitions()
+    {
+        for (String format : List.of("ORC", "PARQUET", "AVRO")) {
+            String hiveTableName = "test_add_files_" + randomNameSuffix();
+            String icebergTableName = "test_add_files_" + randomNameSuffix();
+
+            assertUpdate("CREATE TABLE hive.tpch." + hiveTableName + " WITH (format = '" + format + "') AS SELECT 1 x", 1);
+            assertUpdate("CREATE TABLE iceberg.tpch." + icebergTableName + " WITH (format = '" + format + "') AS SELECT 2 y", 1);
+
+            assertQueryFails(
+                    "CALL iceberg.system.add_files('tpch', '" + icebergTableName + "', 'tpch', '" + hiveTableName + "')",
+                    "Column 'x' does not exist");
+
+            assertUpdate("DROP TABLE hive.tpch." + hiveTableName);
+            assertUpdate("DROP TABLE iceberg.tpch." + icebergTableName);
+        }
+    }
+
+    @Test
+    void testAddFilesDifferentPartitionColumnDefinitions()
+    {
+        String hiveTableName = "test_add_files_" + randomNameSuffix();
+        String icebergTableName = "test_add_files_" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE hive.tpch." + hiveTableName + " WITH (partitioned_by = ARRAY['hive_part']) AS SELECT 1 x, 10 hive_part", 1);
+        assertUpdate("CREATE TABLE iceberg.tpch." + icebergTableName + " WITH (partitioning = ARRAY['iceberg_part']) AS SELECT 2 x, 20 iceberg_part", 1);
+
+        assertQueryFails(
+                "CALL iceberg.system.add_files('tpch', '" + icebergTableName + "', 'tpch', '" + hiveTableName + "')",
+                "Column 'hive_part' does not exist");
+
+        assertUpdate("DROP TABLE hive.tpch." + hiveTableName);
+        assertUpdate("DROP TABLE iceberg.tpch." + icebergTableName);
+    }
+
+    @Test
+    void testAddFilesPartitionFilter()
+    {
+        String hiveTableName = "test_add_files_partition_" + randomNameSuffix();
+        String icebergTableName = "test_add_files_partition_" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE iceberg.tpch." + icebergTableName + "(id int, part varchar) WITH (partitioning = ARRAY['part'])");
+        assertUpdate("CREATE TABLE hive.tpch." + hiveTableName + "(id int, part varchar) WITH (partitioned_by = ARRAY['part'])");
+        assertUpdate("INSERT INTO hive.tpch." + hiveTableName + " VALUES (1, 'test1'), (2, 'test2'), (3, 'test3'), (4, 'test4')", 4);
+
+        assertUpdate("CALL iceberg.system.add_files('tpch', '" + icebergTableName + "', 'tpch', '" + hiveTableName + "', ARRAY['part=test1'])");
```

Comment:
What should happen when there is no partition filter specified for partitioned table? Is `recursive_directory` taken into account?

Reply:
The argument should be used regardless of partitioned or not because the argument is for a storage layer. Sure, I will add a test.

Reply:
> It will result in no-op because of the unmatched partition filter. I will take further look if we should treat it as all partitions.

I would assume that if no filter is specified then all files will be copied as in docs first example, but it will fail for partitioned table. It may also lead to complicated syntax in case table with many partitions.